### PR TITLE
Feat/custom training loop & benchmark API & fix BatchShuffledDataloaderIterator next() call get() twice.

### PR DESCRIPTION
**Pull request by L-M-Sherlock**
@dae, I refactored the training function. Now it will provide the best parameters in all epochs. And I add an extra API for benchmarking. The original one doesn't allow to input the testset, and it's the main difference compared with other models' benchmark processes. I have added the API in my local Anki source code. The latest benchmark results show that FSRS-rs has reached the 99% performance of the Python implementation of the optimizer.
After merging this PR, I will submit the PR for Anki.

I have tested it with collections from real users. It's weird that the progress reach 100% and back to ~80% sometime. Do you know what causes this problem? I'm not familiar with the code of progress bar.

### CODE REVIEW

**Comment by dae** [comment]
This is approximately 4x slower than the previous implementation - I presume that's expected? I didn't spot any obvious issues in the code.

> It's weird that the progress reach 100% and back to ~80% sometime. Do you know what causes this problem?

CombinedProgressState stores Vec<ProgressState>, where is n_splits in length. Each of the splits is calculated in parallel, and when Anki wants to check on the current progress, it calls CombinedProgressState's current() and total(). If progress is going backwards, is it possible that either the total or current figures for one of those splits is decreasing?

**Comment by L-M-Sherlock** [comment]
> This is approximately 4x slower than the previous implementation

The previous implementation employ early stop. For example:

```
[fsrs::training][INFO] epoch: 1 iteration: 0 loss: 0.202727273106575
[fsrs::training][INFO] epoch: 2 iteration: 0 loss: 0.2038394659757614
[fsrs::training][INFO] epoch: 3 iteration: 0 loss: 0.20324841141700745
[fsrs::training][INFO] epoch: 4 iteration: 0 loss: 0.2024657279253006
[fsrs::training][INFO] epoch: 5 iteration: 0 loss: 0.20247885584831238
[fsrs::training][INFO] best_loss: 0.2024657279253006
```

The previous implementation will stop in epoch 2 because the loss increases. The current implementation will finish all epoch. So it's slower than the previous version.

**Comment by dae** [comment]
There seems to be something else going on. On this PR:

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 40 filtered out; finished in 1.85s

On 0.4.6, where early stop was changed:

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 40 filtered out; finished in 0.32s

On 0.4.5: 

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 40 filtered out; finished in 0.25s

**Comment by L-M-Sherlock** [comment]
> test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 40 filtered out; finished in 1.85s

I removed multi_thread() from BatchShuffledDataLoader. Maybe it's the cause. I will add it back and test it again.

**Comment by L-M-Sherlock** [comment]
```
[fsrs::training::tests][INFO] progress: 5120/218000
[fsrs::training::tests][INFO] progress: 12800/218000
[fsrs::training::tests][INFO] progress: 28160/218000
[fsrs::training::tests][INFO] progress: 43600/218000
[fsrs::training::tests][INFO] progress: 43600/218000
[fsrs::training::tests][INFO] progress: 43600/218000
[fsrs::training][INFO] epoch: 1 iteration: 0 loss: 0.20272022485733032
[fsrs::training][INFO] epoch: 1 iteration: 0 loss: 0.20420211553573608
[fsrs::training::tests][INFO] progress: 43600/218000
[fsrs::training][INFO] epoch: 1 iteration: 0 loss: 0.20821613073349
[fsrs::training][INFO] epoch: 1 iteration: 0 loss: 0.20343279838562012
[fsrs::training][INFO] epoch: 1 iteration: 0 loss: 0.2027272880077362
[fsrs::training::tests][INFO] progress: 55888/218000
[fsrs::training::tests][INFO] progress: 67664/218000
[fsrs::training::tests][INFO] progress: 81488/218000
[fsrs::training::tests][INFO] progress: 87200/218000
[fsrs::training::tests][INFO] progress: 87200/218000
[fsrs::training::tests][INFO] progress: 87200/218000
[fsrs::training][INFO] epoch: 2 iteration: 0 loss: 0.2032165229320526
[fsrs::training][INFO] epoch: 2 iteration: 0 loss: 0.20389850437641144
[fsrs::training][INFO] epoch: 2 iteration: 0 loss: 0.2031388282775879
[fsrs::training][INFO] epoch: 2 iteration: 0 loss: 0.20348939299583435
[fsrs::training][INFO] epoch: 2 iteration: 0 loss: 0.2038394808769226
[fsrs::training::tests][INFO] progress: 91808/218000
[fsrs::training::tests][INFO] progress: 111264/218000
[fsrs::training::tests][INFO] progress: 126624/218000
[fsrs::training::tests][INFO] progress: 130800/218000
[fsrs::training::tests][INFO] progress: 130800/218000
[fsrs::training::tests][INFO] progress: 130800/218000
[fsrs::training][INFO] epoch: 3 iteration: 0 loss: 0.20382142066955566
[fsrs::training][INFO] epoch: 3 iteration: 0 loss: 0.2034166008234024
[fsrs::training][INFO] epoch: 3 iteration: 0 loss: 0.2030853033065796
[fsrs::training][INFO] epoch: 3 iteration: 0 loss: 0.20305593311786652
[fsrs::training::tests][INFO] progress: 135920/218000
[fsrs::training][INFO] epoch: 3 iteration: 0 loss: 0.20324841141700745
[fsrs::training::tests][INFO] progress: 147696/218000
[fsrs::training::tests][INFO] progress: 159472/218000
[fsrs::training::tests][INFO] progress: 172832/218000
[fsrs::training::tests][INFO] progress: 174400/218000
[fsrs::training::tests][INFO] progress: 174400/218000
[fsrs::training::tests][INFO] progress: 174400/218000
[fsrs::training][INFO] epoch: 4 iteration: 0 loss: 0.20272360742092133
[fsrs::training][INFO] epoch: 4 iteration: 0 loss: 0.20302720367908478
[fsrs::training][INFO] epoch: 4 iteration: 0 loss: 0.202384352684021
[fsrs::training][INFO] epoch: 4 iteration: 0 loss: 0.20249547064304352
[fsrs::training][INFO] epoch: 4 iteration: 0 loss: 0.20246577262878418
[fsrs::training::tests][INFO] progress: 187200/218000
[fsrs::training::tests][INFO] progress: 198976/218000
[fsrs::training::tests][INFO] progress: 213840/218000
[fsrs::training::tests][INFO] progress: 218000/218000
[fsrs::training::tests][INFO] progress: 218000/218000
[fsrs::training::tests][INFO] progress: 218000/218000
[fsrs::training][INFO] epoch: 5 iteration: 0 loss: 0.20274792611598969
[fsrs::training][INFO] best_loss: 0.20272022485733032
[fsrs::training][INFO] epoch: 5 iteration: 0 loss: 0.20311586558818817
[fsrs::training][INFO] best_loss: 0.20302720367908478
[fsrs::training][INFO] epoch: 5 iteration: 0 loss: 0.20240464806556702
[fsrs::training][INFO] best_loss: 0.202384352684021
[fsrs::training][INFO] epoch: 5 iteration: 0 loss: 0.20245565474033356
[fsrs::training][INFO] best_loss: 0.20245565474033356
[fsrs::training][INFO] epoch: 5 iteration: 0 loss: 0.20247896015644073
[fsrs::training][INFO] best_loss: 0.20246577262878418
[fsrs::training::tests][INFO] progress: 218000/218000
[fsrs::training::tests][INFO] progress: 218000/218000
```

The progress seems normal.

**Comment by L-M-Sherlock** [comment]
> test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 40 filtered out; finished in 1.85s

I forget to add `into_par_iter()` back in the unit test.

**Comment by dae** [comment]
Ok, that's better :-)

**Comment by L-M-Sherlock** [comment]
> I removed multi_thread() from BatchShuffledDataLoader. Maybe it's the cause. I will add it back and test it again.

I failed to do that. The current code of BatchShuffledDataLoader is refactored to avoid this bug:
- https://github.com/tracel-ai/burn/issues/1426

We can add it back until the above issue is resovled.

**Comment by L-M-Sherlock** [comment]
I reduce the batch_size of `dataloader_valid`. Now it's consistent with the previous implementation.

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 40 filtered out; finished in 0.36s

**Comment by dae** [comment]
If that doesn't affect output quality, great!

**Review by asukaminato0721 — APPROVED** [review]
